components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +301,150 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
 
         return oAuthClientAuthenticator.canAuthenticate(request, bodyContentMap, oAuthClientAuthnContext);
     }
+
+    /**
+     * Obtain the client authentication method configured for the application.
+     *
+     * @param clientId     Client ID of the application.
+     * @return Configured client authentication method for the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    public List<String> getConfiguredClientAuthMethods(String clientId) throws OAuthClientAuthnException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+            for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                if (OAuthConstants.TOKEN_ENDPOINT_AUTH_METHOD.equals(serviceProviderProperty.getName())) {
+                    return Arrays.asList(serviceProviderProperty.getValue());
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Error occurred while retrieving the service provider.",
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+
+        return IdentityUtil.getPropertyAsList(FAPI_CLIENT_AUTH_METHOD_CONFIGURATION);
+    }
+
+    /**
+     * Obtain the client ID of the application from the request.
+     *
+     * @param request       Http servlet request.
+     * @return Client ID of the application.
+     * @throws OAuthClientAuthnException OAuth Client Authentication Exception.
+     */
+    private String extractClientId(HttpServletRequest request, Map<String, List> contentParams)
```

Comment:
Since we allow to plugin custom authentication methods as well I feel we should take this resolution part into the individual authenticators and introduce a new method to the interface[1] to resolve the client id. Otherwise if a custom authentication option is developed it will not be usable incase client id resolution also depends on it. We can move all these resolution logic to the individual authentication options which we currently have. And if there is a common one that doesn't fall into the individual option we can add it here.

Reply:
Seems like the interface already has a method for this. https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthenticator.java#L66

Reply:
Here the issue is that we cannot use the methods at the authenticator level as this is where the authenticators to be used for the request are being selected. Therefore we need to get the Client ID to know whether the application is a FAPI compliant application and then the client authenticators are being set according to it.

Reply:
Same as the canAuthenticate method we can loop over the `getClientId` method provided in the interface and resolve the client id. We don't need to engage authentication.

Reply:
Fixed in 130f21ef0e785e76166a4f571458d5d14d43b6ca